Exynos Backend for Executorch to bring up on Exynos SoC (#13677)
pytorch/executorch: backends/samsung/README.md

# ExecuTorch Samsung Exynos Delegate

> The subtree contains Exynos delegation implementation for ExecuTorch. The target of delegation

Reviewer: ```suggestion
The subtree contains Exynos delegate implementation for ExecuTorch. The target of delegation
```

> is deploying torch model run with exynos NPU/DSP.

Reviewer: ```suggestion
is deploying torch model to run with exynos NPU/DSP.
```

Author: Thanks, it's fixed.

This backend is implemented on the top of [EXYNOS_LITECORE](https://soc-developer.semiconductor.samsung.com/global/development/light-core)

> Please prepare the SDK before you start, it is important to code compilation and runtime.

Reviewer: How do we do that? Is there a documentation? Or clear readme?

Ideal path is for us to have exynos_litecore as pip package so that you can just do `pip install exynos_litecore`

Author: we'll consider installing the LiteCore library using pip later.

Reviewer 2: Created an issue to track: https://github.com/pytorch/executorch/issues/14004

## Delegate Options

### Supported Chipset
- Exynos 2500 (E9955)

### Supported Inference Type

> - Quantized (i8/u8/i16/u16)

Reviewer: what about int4? is it under dev?

Reviewer 2: Created an issue to track: https://github.com/pytorch/executorch/issues/14003

- FP16

## Directory Structure

```
backends/samsung
├── aot        # Codes for generating binary buffer for ENN runtime.
├── builders   # Codes for lowering each operators.
├── partition  # ENN Partitioner.
├── passes     # Various passes helping lower models to ENN backend.
├── python     # Places to put pybind artifacts for accessing samsung libraries.
├── runtime    # ENN runtime for executing lowered models.
├── scripts    # Misc supporting scripts, not related to core functionality.
└── serialization # Codes for building Graph IR for Exynos and serializing.

examples
└── samsung # Examples to run ENN backends.
```

## How to build
Please download Exynos AI LiteCore, and set the root path of SDK directory to `EXYNOS_AI_LITECORE_PATH`.</br>
Please navigate to [Android NDK](https://developer.android.com/ndk) and download a version of NDK.
`ANDROID_NDK` refers the root path of NDK directory.</br>

### Set up environment variables
```bash
export LD_LIBRARY_PATH=${EXYNOS_AI_LITECORE_PATH}/lib64
```

### Build AOT Targets

> Generated python artifacts allow user call `Compile` interface to lower a model to ENN backend in python script.

Reviewer: ```suggestion
Generates python artifacts that allow user call `compile` interface to lower a model to ENN backend in python script.
```

Author: Thanks, it's fixed.

```bash

> ./backends/samsung/build.sh -b x86_64

Reviewer 2: Do you wanna add a basic end-to-end integration test to our CI? Can we run everything in x86 environment by any chance?

See example test cases in https://github.com/pytorch/executorch/blob/main/.github/workflows/pull.yml (runs on every PR before merging with main) and https://github.com/pytorch/executorch/blob/main/.github/workflows/trunk.yml (jobs that runs after the PR merges with main)

Author: > > i think you also need to get samsung device for testing CI.
> 
> yeah, we can look into it. also, is there an emulator on linux (either x86 or aarch64) or mac for exynos?

Unfortunately, we don't have the emulator. Could we discuss on slack for this?

Author: > I still think we should add basic unit test, at least for ahead-of-time stuff such as partitioner and serialization. i suppose they won't require samsung phones, right?
> 
> for e2e, we can do it after the PR lands as a fast-follow-on.
you mean we are going to test only until the PTE is generated, right?

Reviewer 2: For e2e integration test, created an issue to track: https://github.com/pytorch/executorch/issues/14002

```

### Build ENN Runtime
```bash
./backends/samsung/build.sh -b android --ndk ${ANDROID_NDK}
```

> ANDROID_ABI=arm64-v8a is default, necessary runtime executable generate in `build_exynos_android` directory.

Reviewer: ```suggestion
ANDROID_ABI=arm64-v8a is default, necessary runtime executable generated in `build_exynos_android` directory.
```

> ANDROID_ABI=arm64-v8a is default, necessary runtime executable generate in `build_exynos_android` directory.

Reviewer: what does it generate

### Build Anroid Extension
This is later exposed Java app. Please turn on CMake option `EXECUTORCH_BUILD_ENN`, and ENN runtime will be added.
```bash
cmake extension/android \
  -DCMAKE_TOOLCHAIN_FILE="${ANDROID_NDK}/build/cmake/android.toolchain.cmake" \
  -DANDROID_ABI="${ANDROID_ABI}" \
  -DCMAKE_INSTALL_PREFIX=cmake-android-out \

>   -Bcmake-android-out/extension/android

Reviewer 2: cc @Jiseong-oh please take a look at Kimish's question

>   -Bcmake-android-out/extension/android

Author: Fixed it.

> # ExecuTorch Samsung Exynos Delegate

Author: Could you more detail explain about this with example? you mean it means that each API used in the Exynos backend needs to be marked as experimental.

> # ExecuTorch Samsung Exynos Delegate

Reviewer 2: > Should we add basic unit test to this Initial PR? If it needs to be added, could you please provide a guide on how to add CI to executorch via email or slack?

Yes, adding basic c++ unit test into executorch would be great.

You can see how other jobs are implemented in https://github.com/pytorch/executorch/blob/main/.github/workflows/pull.yml

You can create a new job, let's say, `unittest-samsung-exynos` and call scripts. There are a few similar examples there.

> # ExecuTorch Samsung Exynos Delegate

Author: > > you mean it means that each API used in the Exynos backend needs to be marked as experimental.
> 
> If it's a user facing api (either python or C++), mark it as experimental.

Got it. we will mark it if some api is facing to user.

> # ExecuTorch Samsung Exynos Delegate

Reviewer 2: > It just add codes for CI now. 

Thanks @Jiseong-oh 

I'm seeing this error in the latest CI run

```
++ awk -F ' ' '{print $2}'
++ tr -d '\n'
+ TORCH_VERSION=2.9.0a0+gite7152ff
+ [[ 2.9.0a0+gite7152ff != *\+\g\i\t* ]]
+ source .ci/scripts/setup-samsung-linux-deps.sh
++ set -ex
++ AI_LITE_CORE_VERSION=0.5.0
++ download_ai_lite_core 0.5.0
++ AI_LITE_CORE_VERSION=0.5.0
++ _exynos_ai_lite_core_url='https://ko.ai-studio-farm.com:26310/api/v1/buckets/lite-core/objects/download?prefix=05/ubuntu2204/exynos-ai-litecore-v0.5.0,tar.gz'
++ _exynos_lite_core_dir=/tmp/exynos_ai_lite_core
++ mkdir -p /tmp/exynos_ai_lite_core
++ _tmp_archive=/tmp/exynos-ai-litecore-v0.5.0.tar.gz
++ curl --silent --show-error --location --fail --retry 3 --output /tmp/exynos-ai-litecore-v0.5.0.tar.gz 'https://ko.ai-studio-farm.com:26310/api/v1/buckets/lite-core/objects/download?prefix=05/ubuntu2204/exynos-ai-litecore-v0.5.0,tar.gz'
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Error: Process completed with exit code 1.
```

it could be that our ubuntu docker image certificates might be old. 

Could you try adding 

```
sudo update-ca-certificates
```

before your curl command?

> # ExecuTorch Samsung Exynos Delegate

Reviewer 2: Also I see that you added python unit tests

Looks like you have to add this line somewhere in test-samsung-models-linux

(It won't be run as part of default unittest job since it is not in pytest.ini)

```
python -m unittest discover backends/samsung/test/ops/ -p "test_*.py"
```

> # ExecuTorch Samsung Exynos Delegate

Author: okay, I add the command before curl command.

> # ExecuTorch Samsung Exynos Delegate

Reviewer 2: Lastly, in your CI job, perhaps validate that this command run successfully.

```
        for model in $models; do
          python -m executorch.examples.samsung.aot_compiler --model_name=$model -c E9955
        done
```

> # ExecuTorch Samsung Exynos Delegate

Reviewer 2: cc @chong-chen01 @Jiseong-oh

> # ExecuTorch Samsung Exynos Delegate

Author: Yes, we will check if that command for test is running successfully. It should be worked well if downloading of LiteCore is works.
